mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,575 @@
+                dict(
+                    with_box_refine=with_box_refine,
+                    as_two_stage=as_two_stage,
+                    num_decoder_layers=decoder_cfg['num_layers']))
+
+        super().__init__(
+            *args, decoder_cfg=decoder_cfg, bbox_head=bbox_head, **kwargs)
+
+    def _init_layers(self) -> None:
+        """Initialize layers except for backbone, neck and bbox_head."""
+        self.positional_encoding = SinePositionalEncoding(
+            **self.positional_encoding_cfg)
+        self.encoder = DeformableDetrTransformerEncoder(**self.encoder_cfg)
+        self.decoder = DeformableDetrTransformerDecoder(**self.decoder_cfg)
+        self.embed_dims = self.encoder.embed_dims
+        if not self.as_two_stage:
+            self.query_embedding = nn.Embedding(self.num_query,
+                                                self.embed_dims * 2)
+
+        num_feats = self.positional_encoding.num_feats
+        assert num_feats * 2 == self.embed_dims, \
+            f'embed_dims should be exactly 2 times of num_feats. ' \
+            f'Found {self.embed_dims} and {num_feats}.'
+
+        self.level_embeds = nn.Parameter(
+            torch.Tensor(self.num_feature_levels, self.embed_dims))
+
+        if self.as_two_stage:
+            self.enc_output = nn.Linear(self.embed_dims, self.embed_dims)
+            self.enc_output_norm = nn.LayerNorm(self.embed_dims)
+            self.pos_trans_fc = nn.Linear(self.embed_dims * 2,
+                                          self.embed_dims * 2)
+            self.pos_trans_norm = nn.LayerNorm(self.embed_dims * 2)
+        else:
+            self.reference_points_fc = nn.Linear(self.embed_dims, 2)
+
+    def init_weights(self) -> None:
+        """Initialize weights."""
+        super().init_weights()  # TODO: should this be in the front ?
+        for p in self.parameters():
+            if p.dim() > 1:
+                nn.init.xavier_uniform_(p)
+        for m in self.modules():
+            if isinstance(m, MultiScaleDeformableAttention):
+                m.init_weights()
+        if not self.as_two_stage:
+            xavier_init(
+                self.reference_points_fc, distribution='uniform', bias=0.)
+        normal_(self.level_embeds)
+
+    def pre_transformer(
+            self,
+            mlvl_feats: Tuple[Tensor],
+            batch_data_samples: OptSampleList = None) -> Tuple[Dict, Dict]:
+        # TODO: Doc
+        """Prepare the inputs of the Transformer.
+
+        Args:
+            mlvl_feats (Tuple[Tensor]): Multi-level features that may have
+                different resolutions, output from neck.
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`.
+                Defaults to None.
+
+        Returns:
+            Dict[str, Tensor]: The inputs of the Transformer.
```

Comment:
Tuple[Dict]:  The first dict contains  the inputs of encoder and the second dict contains the inputs of decoder

Reply:
`encoder_inputs_dict` contains following keys:
    
    - `feat`: comment the shape and the meanings
    - .......
 
`decoder_inputs_dict` contains following keys:
   
    - `memory_mask`